---
output: github_document  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, 
                      #cache = TRUE,
                      fig.path = "./images/")
```

# Introduction to R for RNA-Seq Workshop

<!---

_Even though you can [![hackmd-github-sync-badge](https://hackmd.io/2ArmQGwGT0uUyL5Ehqy0hQ/badge)](https://hackmd.io/2ArmQGwGT0uUyL5Ehqy0hQ), please make changes by editing [this .Rmd file](https://github.com/nih-cfde/training-rstudio-binder/blob/data/GTEx/r4rnaseq-workshop.Rmd)._

--->

**When:** Wednesday,  April 27, 2022, 10 am - 12 pm PDT [workd clock](https://www.timeanddate.com/worldclock/fixedtime.html?msg=+Introduction+to+R+for+RNA-Sequencing+Analysis&iso=20220427T10&p1=3736&ah=2)   
**Where:** [Zoom](https://zoom.us/j/7575820324?pwd=d2UyMEhYZGNiV3kyUFpUL1EwQmthQT09) and [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/nih-cfde/training-rstudio-binder/data?urlpath=rstudio)  
**Instructors:** Dr. Rayna Harris  
**Organizer:** [The Common Fund Data Ecosystem](https://training.nih-cfde.org/)

**Description:** RNA-Sequencing (RNA-Seq) is a popular method for determining the presence and quantity of RNA in biological samples. In this 2 hour workshop, we will use R to explore publicly-available RNA-Seq data from the [Gene Expression Tissue Project (GTEx)](https://gtexportal.org/home/). Attendees will be introduced to the R syntax, variables, functions, packages, and data structures common to RNA-Seq projects. We will use RStudio to import, tidy, transform, and visualize RNA-Seq count data. Attendees will learn tips and tricks for making the processes of data wrangling and data harmonization more manageable. This workshop will not cover cloud-based workflows for processing RNA-seq reads or statistics and modeling because these topics are covered in our [RNA-Seq Concepts](https://osf.io/kj5av/) and [RNA-Seq in the Cloud](https://github.com/nih-cfde/rnaseq-in-the-cloud/blob/stable/rnaseq-workflow.pdf) workshops. Rather, this workshop will focus on general R concepts applied to RNA-Seq data. Familiarity with R is not required but would be useful.

![](https://hackmd.io/_uploads/SkkxxSHeq.png)


<div class="info">

### Learning Objectives 

In this workshop, you will learn how to use R and RStudio to: 

* import and view files commonly associated with RNA-sequencing experiments 
* select variables and observations that are relevant to research questions (tidy)
* create and rename variables (transform)
* join data frames by common variables (harmonize)
* visualize data using bar graphs, scatter plots, and box plots 

You will produce graphs and tables to answer the of the following motivating questions?

* How many RNA-sequencing samples are in the GTEx project? 
* Do you have enough samples to test the effects of sex, age, hardy scale, and their interactions for all tissues? 
* What is the effect of age on gene expression in the heart?
* How is my gene of interest affected by age in the heart and muscle?

</div>


## Introduction 

The book [“R for Data Science”](https://r4ds.had.co.nz/index.html) provides an excellent framework for using data science to turn raw data into understanding, insight, and knowledge. We will use this framework as an outline for this workshop.

**R** is a statistical computing and data visualization programming language. **RStudio** is an integrated development environment, or IDE, for R programming. R and RStudio work on Mac, Linux, and Windows operating systems. The RStudio layout displays lots of useful information and allows us to run R code from a script and view and save outputs all from one interface.

When you start RStudio, you’ll see two key regions in the interface: the console and the output. When working in R, you can type directly into the console, or you can type into a script. Saving commands in a script will make it easier to reproduce. You will learn more as we go along! 


![](https://hackmd.io/_uploads/H1a8-HHx5.png)

For today's lesson, wef will focus on data from the [Gene-Tissue Expression (GTEx) Project](https://commonfund.nih.gov/gtex). GTEx is an ongoing effort to build a comprehensive public resource to study tissue-specific gene expression and regulation. Samples were collected from 54 non-diseased tissue sites across nearly 1000 individuals, primarily for molecular assays including WGS, WES, and RNA-Seq. 

![](https://i.imgur.com/AfjBaPE.png)

![](https://i.imgur.com/XyTGHTK.png)

![](https://i.imgur.com/jvhRnUy.png)



### Getting Started

1. Click the [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/nih-cfde/training-rstudio-binder/data?urlpath=rstudio) button to generate a computing environment for this workshop.
2. Navigate to the GTEx folder.
3. Click `GTEx.Rproj` and click "Yes" to open up an Rproject. This will set the working directory to `~/GTEx/`.
4. If you open the `r4rnaseq-workshop.R` file which contains all the commands for today's workshop, you can click through this and all the commands should run successfully. 
5. If you open a new R Script by clicking **File > New File > R Script**, you can code along by typing out all the commands for today's lesson as I type them.

Click "Run" to send commands from a script to the console or click command enter. 

### R is a calculator

You can perform simple and advanced calculations in R. 

```{r, appendix = TRUE}
2 + 2 * 100
log(202)

```

You can save variable and recall them later.


```{r, appendix = TRUE}
pval <- 0.05
pval

-log10(pval)

```


You can save really long lists of things with a short, descriptive names that are easy to recall later.

```{r, appendix = TRUE}

favorite_genes <- c("BRCA1", "JUN",  "GNRH1", "TH", "AR")
favorite_genes

```

### Loading R packages

Many of the functions we will use are pre-installed. The [Tidyverse](https://www.tidyverse.org/) is a collection of R packages that include functions, data, and documentation that provide more tools and capabilities when using R. You can install the popular data visualization package `ggplot2` with the command  `install.packages("ggplot2")`). It is a good idea to "comment out" this line of code by adding a `#` at the beginning so that you don't re-install the package every time you run the script. For this workshop, the packages listed in the `.binder/environment.yml` file were pre-installed with Conda. 

```{r  eval=FALSE, appendix = TRUE}
#install.packages("ggplot2")

```

After installing packages, we need to load the functions and tools we want to use from the package with the `library()` command. Let's load the `ggplot2` package.

```{r, message=FALSE, appendix = TRUE}
library(ggplot2)
```

<div class="warning">

#### Challenge

We will also use functions from the packages `tidyr` and  `dplyr` to tidy and transform data. What command would you run to load these packages?

<div class="spoiler">

  `library(tidyr)`  
  `library(dplyr)`  
 
</div>


```{r, otherpackages, echo = F, message = F, warning=F , appendix = TRUE}
library(tidyr)
library(dplyr)

```

You can also navigate to the "Packages" tab in the bottom right pane of RStudio to view a list of available packages. Packages with a checked box next to them have been successfully loaded. You can click a box to load installed packages. Clicking the "Help" Tab will provide a quick description of the package and its functions.


<div class="success">

#### Key functions

| Function | Description |
| --- | --- |
| `<-` | The assignment variable | 
| `log10()` | A built-in function for a log transformation  |
| `install.packages()` | An R function to install packages |
| `library()` | The command used to load installed packages |

</div>


## Importing and viewing data

Data can be imported using packages from base R or the tidyverse. What are some differences between the data objects imported by base R functions such as `read.csv()` and Tidyverse functions such as `read_csv()`? To begin with, `read.csv()` replaces spaces and dashes periods in column names, and it also preserves row.names. On the other hand, `read_csv()` preserves spaces and dashes in column names but drops row.names. For this workshop, we will use `read_csv()`, which means we may have to replace dashes with periods so that our sample names in all objects with sample name information.

Today, I will show you how to import the following files:

1. data/samples.csv
1. data/GTExHeart_20-29_vs_70-79.tsv
1. data/colData.HEART.csv
1. data/countData.HEART.csv.gz


Later, you can practice on your own using the following files:

1. data/GTExMuscle_20-29_vs_70-79.tsv
1. data/colData.MUSCLE.csv
1. data/countData.MUSCLE.csv.gz


The `GTExPortal.csv` file in `./data/` contains information about all the samples in the GTEx portal. Let's import this file using `read.csv()`. 

```{r import-1, appendix = TRUE}

samples <- read.csv("./data/samples.csv")

```
After importing a file, there are multiple ways to view the data. `head()` to view the first few lines of each file. `names()` will print just the column names. `str` will compactly displaying the internal structure. `summary` will compute statistics. 

```{r, appendix = TRUE}
View(samples)
head(samples)
tail(samples)
str(samples)
summary(samples)

```




Count files can be very long and wide, so it is a good idea to only view the first (or last) few rows and columns. Typically, a gene identifier (like an ensemble id) will be used as the row names. We can use `dim` to see how many rows and columns are in the file. 


```{r import-4, appendix = TRUE}

counts <- read.csv("./data/countData.HEART.csv", row.names = 1)
dim(counts)
head(counts)[1:5]

```

This "countData" was generated by using `recount3` as described in the file `scripts/recount3.Rmd`. It comes from a Ranged Summarized Experiment (rse) which contains quantitative information about read counts as well as quality control information and sample descriptions. The "colData" from an rse can also be obtained. This information _should_ match the information in our samples file, but there can be subtle differences in formatting We will read the colData in a later section. 


Very large tabular files are often saved as .tsv files. These can be imported with `read.table()` or  `read_tsv()`. You can also specify the tab delimiter as well as the row and column names. You can import files using the default parameters or you can change them. Because the first column in the .tsv files does not have a row name, by default, `read.table()`, imports the first column as the row.names. When `sep = "\t", header = TRUE` is specified, the fist column is imported as column one and given the column name  `X`.

```{r import-2, appendix = TRUE}

# with row.names
results <- read.table("./data/GTEx_Heart_20-29_vs_50-59.tsv")
head(results)

```

<div class="warning">

#### Challenge

What commands could you use to read the following files:
1. GTEx results comparing the muscles of 20-29 year old to 70-79 year olds?
1. The csv file information describing the muscle samples? 


<div class="spoiler">

1. `read.table("./data/GTEx_Muscle_20-29_vs_70-79.tsv")` 
1. `read.csv("./data/countData.MUSCLE.csv", row.names = 1)`  

</div>


#### Quick summary statistics and sample size

You have now seen a variety of options for importing files. You may use many more in your R-based RNA-seq workflow, but these basics will get you started. Let's now explore the functions `summary()`,  `length()`, `dim()`, and `count()` us to quickly summarize and compare data frames to answer the following questions.


How many samples do we have? Over 1400!

```{r, appendix = TRUE}

dim(samples)

```

How many samples are there per tissue? 

```{r, appendix = TRUE}

dplyr::count(samples, SMTS) 

```

How many samples are there per tissue and sex? Can we test the effect of sex on gene expression in all tissues? For many samples, yes, but not all tissues were samples from both males and females. 

```{r, appendix = TRUE}

dplyr::count(samples, SMTS, SEX) 

```

How many samples are there per sex, age, and hardy scale? Do you have enough samples to test the effects of Sex, Age, and Hardy Scale in the Heart? 


```{r, appendix = TRUE}

dplyr::count(samples, SMTS, SEX, AGE, DTHHRDY ) 

```


<div class="warning">

#### Challenge

What series commands would you use to import the `data/colData.MUSCLE.csv` and count the number of muscles samples per sex, age? 

How many female muscles samples are there from age group 30-39?

_Hint: use head() or names() after importing a file to verify the variable names._


<div class="spoiler">

  `df <- read.csv("./data/colData.MUSCLE.csv") `
  `dplyr::count(df, SMTS, SEX, AGE)`
  `# 3 samples are in the female group age 30-39`
  
</div>



<div class="success">

#### Key functions for importing and quickly viewing raw and summarized data

| Function | Description |
| --- | --- |
| `read.csv()`  | A base R function for importing comma separated tabular data  |
| `read_csv()`  | A tidyR function for importing .csv files as tibbles |
| `read.table()` | A base R function for importing tabular data with any delimiter |
| `read_tsv()`  | A tidyR function for importing .tsv files as tibbles | 
| `as_tibble()` | Convert data frames to tibbles | 
| `head()` and `tail()` | Print the first or last 6 lines of an object  | 
| `dim()`  | A function that prints the dimensions of an object | 
| `length()` | Calculate the length of an object |
| `count()` | A dplyr function that counts number of samples per group |
| `str()` | A function that prints the internal structure of an object  |  
| `summary()` | A function that summarizes each variable |  

</div>


## Visualizing data with ggplot2

`ggplot2` is a very popular package for making visualization. It is built on the "grammar of graphics". Any plot can be expressed from the same set of components: a data set, a coordinate system, and a set of "geoms" or the visual representation of data points such as points, bars, line, or boxes. This is the template we build on:


```
ggplot(data = <DATA>, aes(<MAPPINGS>)) +
  <geom_function>() +
  ...
```



We just used the `count()` function to calculate how many samples are in each group. The function for creating bar graphs (`geom_bar()`) also makes use of `stat = "count"` to plot the total number of observations per variable. Let's use ggplot2 to create a visual representation of how many samples there are per tissue, sex, and hardiness.


```{r bar1, appendix = TRUE}

ggplot(samples, aes(x = SMTS)) +
  geom_bar(stat = "count")

```

In the last section, we will discuss how to modify the `themes()` to adjust the axes, legends, and more. For now, let's flip the x and y coordinates so that we can read the sample names. We do this by adding a layer and the function `coord_flip()`

```{r bar2, appendix = TRUE}

ggplot(samples, aes(x = SMTS)) +
  geom_bar(stat = "count") + 
  coord_flip()

```

Now, there are two ways we can visualize another variable in addition to tissue. We can add color or we can add facets.


Let's first color the data by age bracket. Color is an aesthetic, so it must go inside the `aes()`. If you include `aes(color = AGE)` inside `ggplot()`, the color will be applied to every layer in your plot. If you add `aes(color = AGE)` inside `geom_bar()`, it will only be applied to that layer (which is important later when you layer multiple geoms. 

head(samples)

```{r bar3, appendix = TRUE}

ggplot(samples, aes(x = SMTS, color = AGE)) +
  geom_bar(stat = "count") + 
  coord_flip()

```

Note that the bars are outlined in a color according to hardy scale. If instead, you would the bars "filled" with color, use the aesthetic `aes(fill = AGE)`

```{r bar4, appendix = TRUE}

ggplot(samples, aes(x = SMTS, fill = AGE)) +
  geom_bar(stat = "count") + 
  coord_flip()

```

Now, let's use `facet_wrap(~SEX)` to break the data into two groups based on the variable sex.

```{r bar5, appendix = TRUE}

ggplot(samples, aes(x = SMTS, fill = AGE)) +
  geom_bar(stat = "count") + 
  coord_flip() +
  facet_wrap(~SEX)

```

![](https://i.imgur.com/AfjBaPE.png)

With this graph, we have an excellent overview of the total numbers of RNA-Seq samples in the GTEx project, and we can see where we are missing data (for good biological reasons). However, this plot doesn't show us Hardy Scale. It's hard to layer 4 variables, so let's remove Tissue as a variable by focusing just on one Tissue.


<div class="warning">

#### Challenge

Create a plot showing the total number of samples per Sex, Age Bracket, and Hardy Scale for _just_ the Heart samples. Paste the code you used in the chat. 

<div class="spoiler">

There are many options. Here are a few.

```
  ggplot(samples, aes(x = DTHHRDY, fill = AGE))  +
      geom_bar(stat = "count") +
      facet_wrap(~SEX) 
    
  ggplot(samples, aes(x = AGE, fill = as.factor(DTHHRDY)))  +
      geom_bar(stat = "count") +
      facet_wrap(~SEX) 
```      
    
</div>

One thing these plots show us is that we don't have enough samples to test the effects of all our experimental variables (age, sex, tissue, and hardy scale) and their interactions on gene expression. We can, however, focus on one or two variables or groups at a time. 


Earlier, we imported the file "data/GTEx_Heart_20-29_vs_70-79.tsv")" and saved it as "results".  This file contains the results of a differential gene expression analysis comparing heart tissue from 20-29 to heart tissue from 30-39 year olds. This is a one-way design investigating only the effect of age (but not sex or hardy scale) on gene expression in the heart.  Let's visualize these results.

[Volcano Plots](https://en.wikipedia.org/wiki/Volcano_plot_(statistics)) are a type of scatter plots that show the log fold change (logFC) on the x axis and the inverse log (`-log10()`) of a  p-value that has been corrected for multiple hypothesis testing (adj.P.Val). Let's create a Volcano Plot using the `gplot()` and `geom_point()`. _Note: this may take a minute because there are 15,000 points that must be plotted_

```{r, volcano1, appendix = TRUE}

ggplot(results, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point() 

```

The inverse log of p < 05 is `r -log10(0.05)`. We can add a horizontal line to our plot using `geom_hline()` so that we can visually see how many genes or points are significant and how many are not. 

```{r, volcano2, appendix = TRUE}

ggplot(results, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05))

```


```{r, volcano3, appendix = TRUE}

ggplot(results, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point(aes(color = ifelse( adj.P.Val < 0.05, "p < 0.05", "NS"))) +
  geom_hline(yintercept = -log10(0.05)) 

```



```{r, volcano4, appendix = TRUE}

ggplot(results, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point(aes(color = ifelse( adj.P.Val < 0.05, "p < 0.05", "NS"))) +
  geom_hline(yintercept = -log10(0.05))  +
  theme(legend.position = "bottom") +
  labs(color = "20-29 vs 50-59 year olds", 
       subtitle = "Heart Tissue Gene Expression")


```



<div class="warning">

#### Challenge

Create a volcano plot for the results comparing the heart tissue of 20-29 year olds to that of 70-70 year olds? Are there more or less differential expressed gene between 20 and 30 year olds or 20 and 70 year olds?

<div class="spoiler">

```
  df <- read.table("./data/GTEx_Heart_20-29_vs_70-79.tsv")

  ggplot(df, aes(x = logFC, y = -log10(adj.P.Val))) +
    geom_point() +
    geom_hline(yintercept = -log10(0.05))
  
  # more  
```

</div>

In addition to containing information about the donor tissue, the samples file contains has a column with a RIN score, which tells us about the quality of the data. If we wanted to look for interactions between RIN score (SMRIN) and sequencing facility (SMCENTER), we can use a box plot. 

```{r boxplot, appendix = TRUE}

ggplot(samples, aes(x = SMCENTER, y = SMRIN)) +
  geom_boxplot() +
  geom_jitter(aes(color = SMRIN))

```


Now you know a handful of R functions for importing, summarizing, and visualizing data. In the next section, we will tidy and transform our data so that we can make even better summaries and figures. In the last section, you will learn ggplot function for making fancier figures.  

<div class="success">

#### Key functions

| Function | Description |
| --- | --- |
| `ggplot2` | An open-source data visualization package for the statistical programming language R | 
| `ggplot()`  | The function used to construct the initial plot object, and is almost always followed by + to add component to the plot |
| `aes()`  | Aesthetic mappings that describe how variables in the data are mapped to visual properties (aesthetics) of geoms |
| `geom_point()`  | A function used to create scatter plots | 
| `geom_bar()`  | A function used to create bar plots | 
| `coord_flip()`  | Flips the x and y axis | 
| `geom_hline()`  | Add a horizontal line to plots | 

</div>

## Tidy and Transform Data

[Data wrangling](https://en.wikipedia.org/wiki/Data_wrangling) is the process of tidying and transforming data to make it more appropriate and valuable for a variety of downstream purposes such as analytics. The goal of data wrangling is to assure quality and useful data. Data analysts typically spend the majority of their time in the process of data wrangling compared to the actual analysis of the data.

**Tidying** your data means storing it in a consistent form. When your data is tidy, each column is a variable, and each row is an observation. Tidy data is important because the consistent structure lets you focus your struggle on questions about the data, not fighting to get the data into the right form for different functions. Some tidying functions include `pivot_longer()`, `pivot_wider()`, `separate()`, `unite()`, `drop_na()`, `replace_na()`. The "lubridate" package has a number of functions for tidying dates. You may also use `mutate()` function to convert objects from, say, characters or integers to factors or rename observations and variables.


**Transforming** your data includes narrowing in on observations of interest (like all people in one city, or all data from the last year), creating new variables that are functions of existing variables (like computing speed from distance and time), and calculating a set of summary statistics (like counts or means). Summary functions such as `summarize()` and `count()` to create new tables with statistics. Before summarizing or counting a whole data frame, you can use `group_by()`  to group variables.  You can use `filter()` and `select()` to isolate specific rows or columns, respectively. If you want to sort columns, `arrange()` and `arrange(desc())` are two functions to familiarize yourself with.

**Combining tables** can be accomplished in one of two ways. If all the columns or  all the rows have all the same names, you can use `rbind()` or `cbind()`, respectively, to join the data frames. If however, each data frame have a column (or multiple columns) that contain unique identifiers, then you can use the family of join functions (`inner_join()`, `outter_join()`, `left_join()`, and `right_join()`)


For each downstream analysis, you will likely use a series of tidying and transforming steps in various order to get your data in the appropriate format. Interest of creating dozens of intermediate files after each step, we will use the `%>%` operator to "pipe" the output of one function to the input of the other. 

Instead of going into each function or each process in detail in isolation, let's start with some typical research questions and then piece together R functions to get the desired information

###  Filtering Data

Filter is done in a few different ways depending on the type of variable. You can use `>` and less `<` to filter greater or less than a number. `==` and `!=` are used to filter by characters or factors that match or do not match a specific pattern. `%in% c()` is used to filter by things in a list. Let's filter by adjusted p-value. You can use `|` and `&` to mean "or" or "and"

To explore filtering data, let's answer the following question: What are the approved names and symbols of the differentially expressed genes (DEGs) in the heart tissue between 20-29 and 30-29 year olds? To answer this question, we need a subset of information from both the results and genes files.  We need, in no particular order, to:

1. filter by adj.p.value  < 0.05 (or desired alpha) 
1. filter by results by logFC > 1 or <-1
1. filter by a list of gene symbols


```{r wrangle11, appendix = TRUE}

results %>% filter(adj.P.Val < 0.05) %>% head()

```


```{r wrangle12, appendix = TRUE}

results %>% filter(logFC > 1 | logFC < -1) %>% head()

```

Sometimes its nice to arrange by pvalue.

```{r wrangle13a, appendix = TRUE}

results %>% filter(adj.P.Val < 0.05,
                   logFC > 1 | logFC < -1) %>%
  arrange(adj.P.Val) %>%
  head()


```


```{r wranglelist, appendix = TRUE}

resultsDEGs <- results %>% filter(adj.P.Val < 0.05,
                   logFC > 1 | logFC < -1) %>%
  arrange(adj.P.Val) %>% 
  rownames(.)
resultsDEGs

```

<div class="warning">

#### Challenge

Replace the input results file with a different file, such as the results of the comparison of 20-29 and 50-59 year old heart samples. What are the deferentially expressed genes?

<div class="spoiler">

You could use the following code to get this result below

```

resultsDEGs2 <- read.table("./data/GTEx_Heart_20-29_vs_50-59.tsv") %>% 
  filter(adj.P.Val < 0.05,
                   logFC > 1 | logFC < -1) %>%
  arrange(adj.P.Val) %>% 
  rownames(.)
resultsDEGs2

[1] "EDA2R"        "PTCHD4"       "BTBD11"       "MTHFD2P1"     "C4orf54"      "LOC101929331"
[7] "FMO3"         "KLHL41"       "ETNPPL"       "HOPX"         "PDIA2"        "RPL10P7"     
[13] "FCMR"         "RAD9B"        "LMO3"         "NXF3"         "FHL1"         "EREG"        
[19] "CHMP1B2P"     "MYPN"         "VIT"          "XIRP1"        "DNASE1L3"     "LIPH"        
[25] "PRELP"        "CSRP3"        "FZD10-AS1"    "LINC02268"    "GDF15"        "PHF21B"      
[31] "CPXM1"        "IL24"         "ADH1B"        "MCF2"         "WWC1"         "SGPP2"       
[37] "COL24A1"      "SEC24AP1"     "ANKRD1"       "CDO1"         "CCL28"        "SLC5A10"     
[43] "XIRP2" 

```

</div>






###  Mutating Data



Most RNA-Seq pipelines require that the counts file to be in a matrix format where each sample is a column and each gene is a row and all the values are integers or doubles with all the experimental factors in a separate file. More over, we need a corresponding file where the row names are the sample id and they match the column names of the counts file. 

When you type `rownames(colData) == colnames(counts)` you should see many TRUE statments. If the answer if FALSE your data cannot be processed by downstream tools. 


```{r question2a, appendix = TRUE}

colData <- read.csv("./data/colData.HEART.csv", row.names = 1)
head(colData)

head(rownames(colData) == colnames(counts))
head(colnames(counts))
head(rownames(colData))

```

The row and col names don't match because the the dashes were replaced with periods when the data were imported. This is kind of okay because `DESeq2` would complain if your colnames had dashes. We can use `gsub()` to replace the dashes with periods. 



```{r question2b, appendix = TRUE}

colData_tidy <-  colData %>%
  mutate(SAMPID = gsub("-", ".", SAMPID))  
rownames(colData_tidy) <- colData_tidy$SAMPID

mycols <- rownames(colData_tidy)
head(mycols)

```

Then, we rename the row names. We can use `select(all_of())` to make sure that all the rows in colData are represented at columns in countData. We could modify the original files, but since they are so large and importing taking a long time, I like to save "tidy" versions for downstream analyses. 



```{r question2b2, appendix = TRUE}

counts_tidy <- counts %>%
  select(all_of(mycols))

head(rownames(colData_tidy) == colnames(counts_tidy))

```


### Joining Data 

Genes can be identified by their name, their symbol, an Ensemble ID, or any number of other identifiers. Our results file uses gene symbols, but our counts file uses Ensemble IDs. 

Let's read a file called "genes.txt" and combine this with our results file so that we have gene symbols, names, and ids, alongside with the p-values and other statistics.


```{r import-3, appendix = TRUE}

genes <- read.table("./data/genes.txt", sep = "\t",  header = T, fill = T)
head(genes)

```

The column with genes symbols is called `Approved.symbol`. To combine this data frame without results. We can use the mutate function to create a new column based off the row names. Let's save this as `resultsSymbol`.


```{r addcoumn, appendix = TRUE}

resultsSymbol <- results %>%
  mutate(Approved.symbol = row.names(.))
head(resultsSymbol)

```

Now, we can use one of the join functions to combine two data frames. `left_join` will return all records from the left table and any matching values from the right. `right_join` will return all values from the right table and any matching values from the left. `inner_join` will return records that have values in both tables. `full_join` will return everything. 

```{r appendix = TRUE}

resultsName <- left_join(resultsSymbol, genes, by = "Approved.symbol")
head(resultsName)

```


Congratulations! You have successfully joined two tables. Now, you can filter and select columsn to make a pretty table of the DEGS.

. 
```{r appendix = TRUE}

resultsNameTidy <- resultsName %>%
  filter(adj.P.Val < 0.05,
                   logFC > 1 | logFC < -1) %>%
  arrange(adj.P.Val) %>%
  select(Approved.symbol, Approved.name, Ensembl.gene.ID, logFC, AveExpr, adj.P.Val)
head(resultsNameTidy)

```




### Lengthen data

The matrix form of the count data is required for some pipelines, but many R programs are better suited to data in a long format where each row is an observation. I like to create  `counts_tidy_long` file that can be easily subset by variables or genes of interest. 

Because the count files are so large, it is good to filter the counts first. I'll filter by `rowSums(.) > 0` and then take the top 6 with `head()`. Then crate a column for lengthening.




```{r question2c, appendix = TRUE}

counts_tidy_slim <- counts_tidy %>%
  filter(rowSums(.) >0 ) %>%
  head() %>%
  mutate(Ensembl.gene.ID = row.names(.) )
head(counts_tidy_slim)
```

Now we can pivot longer. We use `cols` to specify with column names will be turned into observations and we use `names_to` to specify the name of the new column that contains those observations. We use `values_to` to name the column with the corresonding value, in this case we will call the new columns, `SAMPID` and `counts`.  

```{r question2cb, appendix = TRUE}

counts_tidy_long <- counts_tidy_slim %>%
  pivot_longer(cols = all_of(mycols), names_to = "SAMPID", 
               values_to = "counts") 
head(counts_tidy_long)
```

Now, that we have a `SAMPID` column, we can join this with our colData_tidy and then make some pretty plots.

```{r}
counts_tidy_long_joined <- counts_tidy_long%>%
  inner_join(., colData_tidy, by = "SAMPID") %>%
  arrange(desc(counts))
head(counts_tidy_long_joined)
```




```{r boxplot2, appendix = TRUE}

library(scales)

counts_tidy_long_joined %>%
  ggplot(aes(x = AGE, y = counts)) +
  geom_boxplot() +
  geom_point() +
  facet_wrap(~Ensembl.gene.ID, scales = "free_y") +
  scale_y_log10(labels = label_number_si())

```

That completes our section on tidying and transforming data.

<div class="success">

#### Key functions: Tidy and Transform

| Function         | Description |
|------------------|-------------|
| `filter()`       |  A function for filtering data |
| `mutate()`       |  A function for create new columns  |
| `select()`       |  A function for selecting/reordering columns   |
| `arrange()`      |  A function for ordering observations  |
| `full_join()`    |  Join 2 tables, return all observations |
| `left_join()`    |  Join 2 tables, return all observations in the left and matching observations in the right table |
| `inner_join()`   |  Join 2 tables, return observations with values in both tables        |
| `pivot_wider()`  |  Widen a data frame  |
| `pivot_longer()` |  Lengthen a data frame |
| `drop_na()`      |  Remove missing values  |
| `separate()`     |  Separate a column into two columns   |

</div>


## Communicate

### R Markdown

The workshop notes for using this repository to teach an Introduction to R for RNA-seq are crated with the file `r4rnaseq-workshop.Rmd`. 

### References

- [R for Data Science by Hadley Wickham and Garrett Grolemund](https://r4ds.had.co.nz/index.html)
- [Rouillard et al. 2016. The Harmonizome: a collection of processed datasets gathered to serve and mine knowledge about genes and proteins. Database (Oxford).](http://database.oxfordjournals.org/content/2016/baw100.short)

### Additional Resources

- [RStudio cheatsheet for readr](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-import.pdf)
- [RStudio cheatsheet for dplyr](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-transformation.pdf)
- [RStudio cheatsheet for data Wrangling with dplyr](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf)
- [ggplot point shapes](http://www.sthda.com/english/wiki/ggplot2-point-shapes)
- [Angus 2019 Intro to R Lesson](https://angus.readthedocs.io/en/2019/R_Intro_Lesson.html)
- [Angus 2019 Differential Gene Expression in R Lesson](https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html)
- [Software Carpentry R Lesson](http://swcarpentry.github.io/r-novice-inflammation/)


*Note: the source document [r4rnaseq-workshop.Rmd](https://github.com/nih-cfde/training-rstudio-binder/blob/data/GTEx/r4rnaseq-workshop.Rmd) was last modified 14 March, 2022.*

---


### Appendix

```{r all-code, ref.label = knitr::all_labels(appendix == TRUE), eval = F, indent = ""}

```



